### DISCUSSION TITLE
How can I patch a plugin registered through a setuptools entry point

### DISCUSSION THREAD

**Discussion by dgard1981** [pytest-dev/pytest#11297] [Q&A]
I'm developing a small plugin for use internally that will be registered through a setuptools entry point.

In my attempts to write tests for the plugin, I wish to patch a method that is part of an object included in one of the modules. However, patching directly doesn't have any affect. This isn't really surprising, but I'm unsure of *exactly* how the plugin is loaded and what I would need to patch. I suspect there is something in the `pytestconfig` object that I can patch.

```python
monkeypatch.setattr(MyPluginObject, '_get_node_desc', lambda: 'mocked node description')
```

Is it possible to patch a plugin that is installed via a `setup.py` file?

**Comment by RonnyPfannschmidt** [comment]
the plugin-manager provides mechanisms to access plugin objects

without a clear idea of how your plugin self-enables is hard to give further suggestions

if you self-register a plugin object, i'd recommend to put a reference into the stash using a stash-key to ease finding it

**Reply by dgard1981** [reply]
Thanks for the reply. I'll take a look at the stash to see how I can use that. I am working on making the question better too.

**Reply by dgard1981** [reply]
@RonnyPfannschmidt, the `stash` looks to be the way I need to go. I had to make a slight change to my plugin, so now the value that I want to patch is assigned to an instance parameter and consumed from there, rather than directly from the return value of the method.

The following allows me to patch the stash, but the `monkeypatch.undo()`  method isn't reverting the stash back to how it should at the end of the test.

```python
    @pytest.fixture(scope='function', name='mocknodedesc')
    def mock_node_desc_fixture(self, 
            pytestconfig: pytest.Config,
            monkeypatch: pytest.MonkeyPatch) -> None:

        patched_stash = pytestconfig.stash
        patched_stash['pytest_myplugin'].node_desc = 'mocked node description'
        monkeypatch.setattr(pytestconfig, 'stash', patched_stash)

        yield

        monkeypatch.undo()
```

My understanding is that `monkeypatch.undo()` shouldn't be required, but even explicitly including it doesn't undo the change.

The [documentation](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#yield-fixtures-recommended) for `yield` fixtures states -
> Once the test is finished, pytest will go back down the list of fixtures, but in the reverse order, taking each one that yielded, and running the code inside it that was after the yield statement.

However, it's seems like the code after yield is only run on module or session completion, not function completion.

**Reply by RonnyPfannschmidt** [reply]
there is a key error in your code, instead of monkey-patching the actual object you own, you monkey-patch the config object while permanently changing your plugin object

**Reply by dgard1981** [reply]
@RonnyPfannschmidt, do you mean this code is actually results in `pytestconfig.stash` being permanantly updated because `patched_stash` is only a shallow copy.

```python
patched_stash = pytestconfig.stash
patched_stash['pytest_myplugin'].node_desc = 'mocked node description'
```

Or do you mean that I shouldn't be patching `pytestconfig.stash` at all?

In case it's the first, I have tried to patch the stash directly but that didn't work.

```python
monkeypatch.setattr(pytestconfig.stash['pytest_myplugin'], 'node_desc', 'mocked node description')
```

**Reply by RonnyPfannschmidt** [reply]
you shouldnt be patching stash, you should only patch your plugin object

**Reply by dgard1981** [reply]
Right OK, in that case I have misunderstood your first response where you mentioned the stash. And it means that my original question stands...

Patching the object in my module containing the doesn't work, and I'm guessing that is because there is a copy of the object somewhere in `pytestconfig`. However, I don't know how to find that. All the stash does is give visibility of the instance parameter values on the object, rather than tell me where the object actually is.

**Reply by RonnyPfannschmidt** [reply]
Its not quite clear what you are asking for, if you have a reference, your know where the object is

As there's literally zero extra context provided as of now it's not possible to give useful suggestions

**Reply by dgard1981** [reply]
@RonnyPfannschmidt, So as an example, say I wanted to patch something in pytest-cov. If I wanted to mock the pytest_terminal_summary method of the CovPlugin class from /home/user/.local/lib/python3.10/site-packages/pytest_cov/plugin.py, how would I do that?

In this example lets say that I want to just print Hello World instead of actually printing any of the coverage report.

Note that I don't actually want to do this, but the method that I want to mock is a helper method called by pytest_terminal_summary in my own plugin.py::MyPlugin class, so this is a pretty close and hopefully transferable example.

**Reply by RonnyPfannschmidt** [reply]
The actual hook is not something that's mockable, so the example can't transfer

But if you have a plugin object whose hook calls another method,that method can be monkey patched

**Reply by RonnyPfannschmidt** [reply]
However without More details it's hard to figure lifecycles and more,

**Reply by dgard1981** [reply]
@RonnyPfannschmidt, hopefully this example plugin that you can create in a docker container will demonstrate what I wish to do.

The plugin simply adds a `pytest_terminal_summary` hook that calls out to the method `_some_info`. What I wish to do is patch the return value of `_some_info`. I've added a short explanation of why I want to do this  below, if it helps.

My attempts to do so thus far have failed, including in this example where I import `MyPlugin` and use `monkeypatch.setattr(MyPlugin, '_some_info', lambda: 'foobar')`.

```shell
# Run in a docker container.
docker run --rm -it python:3.10-bullseye /bin/bash

# Make the require directories and change to the plugin project root.
mkdir -p /tmp/example_project/src/test
mkdir -p /tmp/example_project/src/pytest_example
cd tmp/example_project

# Make the plugin file.
cat <<EOF > /tmp/example_project/src/pytest_example/plugin.py
import pytest

class MyPlugin:

    def _some_info(self):
        return 'Some info.'

    @pytest.hookimpl
    def pytest_terminal_summary(self, *_) -> None:
        '''Print details of passed assertions in the terminal summary.'''
        print(self._some_info())

@pytest.hookimpl
def pytest_configure(config: pytest.Config) -> None:
    '''Register the pytest-expect plugin.'''

    config.pluginmanager.register(MyPlugin(), '_example')
EOF

# Make the test file.
cat <<EOF > /tmp/example_project/src/test/test_something.py
from pytest_example.plugin import MyPlugin

def test_something(monkeypatch):

    monkeypatch.setattr(MyPlugin, '_some_info', lambda: 'foobar')
    assert 1 == 1
EOF

# Make the project TOML file.
cat <<EOF > /tmp/example_project/pyproject.toml
[build-system]
requires = ["setuptools"]
build-backend = "setuptools.build_meta"

[project]
name = "pytest_example"
version = "0.0.1"

[project.entry-points.pytest11]
pytest_example = "pytest_example.plugin"
EOF

# Install pytest and the plugin, then run pytest.
pip install pytest /tmp/example_project/
pytest
```

#### Why I'm doing this

In short, my plugin adds output to the terminal summary, and I need to test that the terminal summary includes the expected output. Most of it is predictable, but one line depends on the environment in which the test run occurs, hence the need to mock that part.

The only way I can think to do this is run `pytest.main()` from within a test, and then use the `capsys` fixture to check for the expected output in the terminal summary.

The tests are configured in such a way that running `pytest` only triggers the main test, and the that test passes a specific module to `pytest.main()`, ensuring the tests aren't caught in a never ending loop.

**Reply by RonnyPfannschmidt** [reply]
As far as I can tell,you need the pytester Plugin to run your testing

If you monkeypatch inside a test,it gets undone After the test function

**Reply by dgard1981** [reply]
OK, I see what you mean - the mock is undone before the `pytest_terminal_summary` is fired, so that's why it seems that `monkeypatch` isn't working.

I will take a look at `pytester` (not come across it before) and see if I can figure that out. But from the description, sounds hopeful that it's the right plugin for the job.

**Reply by The-Compiler** [reply]
Yep, `pytester` is definitely the right way to test pytest plugins. Also see the relevant docs: [Writing plugins — pytest documentation](https://docs.pytest.org/en/7.4.x/how-to/writing_plugins.html#testing-plugins)

**Reply by dgard1981** [reply]
@The-Compiler, @RonnyPfannschmidt, you are correct, `pyteter` is the best tool for the job. However, I'm still stuck on my initial point... How do I mock a method in a plugin class?

Take the example below. It will print `Some info.` in the terminal summary (the `pytest_terminal_summary` hook calls out to another method in the same class to get that text). In reality, the value returned by `MyPlugin._some_info()` is unknown (it's to do with the where the tests are run), and that's why I want to mock it to make the output.

So how could I amend this example to mock the return value of `MyPlugin._some_info()` in tests run by `pytester`?

```shell
# Run in a docker container.
docker run --rm -it python:3.10-bullseye /bin/bash


# Make the require directories and change to the plugin project root.
mkdir -p /tmp/example_project/test
mkdir -p /tmp/example_project/src/pytest_example
cd tmp/example_project/test


# Make the plugin file.
cat <<EOF > /tmp/example_project/src/pytest_example/plugin.py
import pytest

class MyPlugin:

    def _some_info(self):
        return 'Some info.'

    @pytest.hookimpl
    def pytest_terminal_summary(self, config, *_) -> None:
        print(self._some_info())

@pytest.hookimpl
def pytest_configure(config: pytest.Config) -> None:
    config.pluginmanager.register(MyPlugin(), '_example')
EOF


# Make the test file.
cat <<EOF > /tmp/example_project/test/test_something.py
import pytest

def test_something(pytester: pytest.Pytester):

    result = pytester.runpytest('-q')
    assert result.stdout.fnmatch_lines('Some info.') is None
EOF


# Make the test INI.
cat <<EOF > /tmp/example_project/test/pytest.ini
[pytest]
addopts = -p no:cacheprovider -p no:example -p pytester
EOF


# Make the project TOML file.
cat <<EOF > /tmp/example_project/pyproject.toml
[build-system]
requires = ["setuptools"]
build-backend = "setuptools.build_meta"

[project]
name = "pytest_example"
version = "0.0.1"

[project.entry-points.pytest11]
pytest_example = "pytest_example.plugin"
EOF


# Install pytest and the plugin, then run pytest.
pip install pytest /tmp/example_project/
pytest 
```
